docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If not defined,
+  // this NetworkPolicyPort does not restrict based on port number.
+  // If defined, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicySource struct {
+  // If 'Namespaces' is defined, 'Pods' must not be.
+  // This is a label selector which selects Pods in this namespace.
+  // This NetworkPolicySource matches any pods selected by this selector.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // If 'Pods' is defined, 'Namespces' must not be.
+  // Selects Kubernetes Namespaces.  This NetworkPolicySource matches 
+  // all pods in all namespaces selected by this label selector. 
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+### Behavior
+
+The following pseudo-code attempts to define when traffic is allowed to a given pod when using this API.
+
+```python
+def is_traffic_allowed(traffic, pod):
+  """
+  Returns True if traffic is allowed to this pod, False otherwise.
+  """
+  if namespace.network_isolation == False:
+    # If network-isolation is disabled, all traffic is allowed.
+    return True 
+  elif traffic.source == pod.node:
+    # Traffic is from the pod's host - this allows for kubelet health checks.
+    return True
+  else:
+    # If network-isolation is enabled, only allow traffic 
+    # that matches a network policy which selects this pod.
+    for network_policy in all_network_policies:
+      if not network_policy.podSelector.selects(pod):
+        # This policy doesn't select this pod. Try the next one. 
+        continue
+
+      # This policy selects this pod.  Check each ingress rule 
+      # defined on this policy to see if it allows the traffic.
+      for ingress_rule in network_policy.ingress:
+        if ingress_rule.from.matches(traffic) and \
+             ingress_rule.ports.matches(traffic):
+          return True 
+
+  # network-isolation is True and no policies match the given pod and traffic.
+  return False
+```
+
+### Open Questions
+
+> The question then remains whether one NP object per "app" makes sense (in this case app={a,b,c}) or whether one NP object per selector
+
+- A single podSelector per NetworkPolicy may lead to managing a large number of NetworkPolicy objects, each of which is small and easy to understand on its own.  Allowing a podSelecor per ingress rule would allow for fewer, larger NetworkPolicy objects that may be easier to manage, but harder to understand individually.  This proposal has opted to favor a larger number of smaller objects that are easier to understand.  Is this the right decision? 
+
+- Is the `Namespaces` selector in the `NetworkPolicySource` struct too coarse? Do we need to support the AND combination of `Namespaces` and `Pods`? 
```

Comment:
> I am not familiar with how the world uses Namespaces. Does it even use them at all?

I think the most common use cases here are for separating applications so that they don't conflict (e.g two apps with the service "db") or for splitting out different users of the cluster.  That said, I don't think there is a prescriptive "this is what namespaces must be used for" type statement out there (correct me if I'm wrong!).  I agree with you that my first inclination is to prevent one namespace from seeing the labels of another.  If this is something users need, we will get that feedback and can modify the behavior accordingly.


If neither `Pods` nor `Namespaces` is defined, the NetworkPolicySource matches all sources.  However, the traffic still needs to match any `NetworkPolicyPorts` that been defined as well, so it isn't a blanket "allow".

The way to express "allow all" in this proposal is to have a `NetworkPolicyIngressRule` with no `Ports` and no `NetworkPolicySource`.

In yaml, that would be:
```yaml
ingress:
  - {}
```